Update: space-in-parens linting messages and report behaviour

GrayedFox opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

- [x] Documentation update
- [x] Changes an existing rule ([template]https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
- [x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))

The current space-in-parens description is a little unclear. Technically empty parens can trigger an error if specified in the exceptions options. This PR updates that line in the docs and adds a paragraph that attempts to explain what exactly this rule does. [Current rule description](https://eslint.org/docs/rules/space-in-parens#rule-details).

The linting error messages are now more specific and inform whether the paren token is either rejecting an extra space or missing a required space. Previously the output did not specify if the space was required or missing (so it could be either).

The previous context/report behavior would sometimes report only the first error encountered within a node due to the token checks being nested inside `else if` blocks. This is a non-breaking change: it simply means all errors will be reported within a node instead of only one - it does not introduce a new error type. The fallout of this change is that there is now at least one interesting case: using the default settings `foo( )` technically contains two errors: that single space is both a rejected opening and closing space, so two errors will be reported despite the fact that removing the one space fixes both errors.

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

GrayedFox commented:
I will follow the contributing guide and rebase and fix my commits messages once this is ready :smile: 

I've added a valid code sample and fixed the previous error that was thrown for the closing paren, but I'm now trying to ensure that given the example rule (see above) we see a rejected space error given this input: `((1+2) )` (this should be rejected due to the trailing space between the closing parens).

The output, strangely enough, is:

```
  2) space-in-parens
       invalid
         ((1 + 2) ):

      AssertionError [ERR_ASSERTION]: 'There must be a space inside this paren.' === 'There should be no spaces inside this paren.'
      + expected - actual

      -There must be a space inside this paren.
      +There should be no spaces inside this paren.
      
      at assertMessageMatches (lib/testers/rule-tester.js:439:24)
      at testInvalidTemplate (lib/testers/rule-tester.js:482:25)
      at Context.RuleTester.it (lib/testers/rule-tester.js:592:25)
```

platinumazure commented:
Hi @GrayedFox, thanks for the PR! Apologies for letting this slip through the cracks.

I've gone through the comments you left in the issue after my latest comment there and now I have a decent idea what's going on here. I'll try to review this tonight and will let you know if I have any questions here. Thanks so much for your patience!

GrayedFox commented:
Hey there, no worries! I've started doing a relatively significant refactor of this rule, since I found the code incredibly hard to read and follow (although it's making more sense now). The last push I did breaks about 50 tests so it's still very much a WIP! Will likely have time to do a bit more work on it tonight/this week

GrayedFox commented:
Hi there! I do, and thanks for the patience - had a nice holiday period and am still wanting to complete this, very much.

Can do a little more work tonight actually just to get the ball rolling again. At the bare minimum, this PR would act as a clean up to the current space-in-parens logic and output more helpful messages, but I'm still aiming to fix what I think is buggy behaviour.

Thanks for checking in :wink:

GrayedFox commented:
Cleaned up the PR description and removed my old out of date comments - it's taken me a long time to properly wrap my head around what I think the bugged behaviour is - and even longer to really understand the internals of the space-in-parens file and tests (it's a jungle out there!).

This PR now has a clear goal in mind and outlines intended changes. Thanks for your patience!

platinumazure commented:
Oh, but this will probably need to be rebased and the commit message fixed.

Commit message should start with "Update:" and should explain what is fixed as briefly as possible. You can add more info later in the commit message (outside the first line) if needed. Thanks!

GrayedFox commented:
Ahoy!

My friend no apologies needed - I have been terribly slack - this is still
actually a work in progress. I'm going to limit the scope of this to just
format the messages to be more descriptive/for clarity. Original goal was
to fix what I thought was a simple bug and then... Enter rabbit hole. Will
rebase and update this afternoon, thanks for the reminder!

On Sun., 26 May 2019, 4:29 am Kevin Partington, <notifications@github.com>
wrote:

> Oh, but this will probably need to be rebased and the commit message fixed.
>
> Commit message should start with "Update:" and should explain what is
> fixed as briefly as possible. You can add more info later in the commit
> message (outside the first line) if needed. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/11121?email_source=notifications&email_token=ABRRI6YZHT5C2SJLTQKRJQLPXHYZTA5CNFSM4GGGE3M2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWH4XMY#issuecomment-495963059>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRI67KXPI3PVCJZGZGGVTPXHYZTANCNFSM4GGGE3MQ>
> .
>

platinumazure commented:
Sounds good @GrayedFox, ping me when you're ready for me to take another look!

GrayedFox commented:
Finally got around to finishing this. Updated docs to follow suit (this does have breaking changes, in that `()` is no longer always allowed). This was a necessary result of getting the rule to properly enforce consistent spacing, even with outer most parentheses, but it comes with another advantage: the "empty" exception now functions as exactly as described (by actually allowing or disallowing empty parens).

GrayedFox commented:
@platinumazure hiya, this thing is finally ready for review again. Docs and rule logic updated to now respect outer most parentheses (fixes #10905).

The main change to the rule logic to achieve this was to check that the token used for checking exceptions is in fact a punctuator (opening or closing paren/bracket/brace) since, it turns out, that in some cases we were passing an Identifier instead.

platinumazure commented on lib/rules/space-in-parens.js:
```diff
@@ -115,38 +118,30 @@ module.exports = {
          */
-        function shouldCloserHaveSpace(left, right) {
-            if (astUtils.isOpeningParenToken(left)) {
+        function openerMissingSpace(left, right) {
```
While we're in this space, could we improve the parameter names? Names like `openingParenToken` and `tokenAfterOpeningParen` would be much easier to read than `left` and `right`.

GrayedFox replied:
Yup, makes sense

platinumazure commented on lib/rules/space-in-parens.js:
```diff
@@ -115,38 +118,30 @@ module.exports = {
-        function shouldOpenerHaveSpace(left, right) {
-            if (sourceCode.isSpaceBetweenTokens(left, right)) {
-                return false;
+        function checkPunctuator(first, second) {
```
Is this function needed? Would it make more sense to use an Array.prototype method inline with a filter function at calling sites (e.g., `const punctuator = [first, second].find(token => token.type === "Punctuator")`?

I'm not sure the name `checkPunctuator` is very clear for this function, in any case. Maybe something like `getFirstPunctuator` would make more sense?

GrayedFox replied:
I've updated the function to instead accept an array of tokens as the only argument and return the  first encountered token that is a Punctuator, but I would be in favour of keeping the function since this exact same inline behavior would be repeated 9 times without the function call and would quite subtly change the order of tokens each time, which I hope is made more obvious by having a function call instead.

Thoughts?

GrayedFox replied:
Removed after further consideration

platinumazure commented on lib/rules/space-in-parens.js:
```diff
@@ -155,24 +150,23 @@ module.exports = {
          */
-        function shouldOpenerRejectSpace(left, right) {
-            if (right.type === "Line") {
+        function openerRejectsSpace(left, right) {
```
While we're in this space, could we improve the parameter names? Names like `openingParenToken` and `tokenAfterOpeningParen` would be much easier to read than `left` and `right`.

platinumazure commented on lib/rules/space-in-parens.js:
```diff
@@ -181,23 +175,35 @@ module.exports = {
          */
-        function shouldCloserRejectSpace(left, right) {
-            if (astUtils.isOpeningParenToken(left)) {
+        function closerMissingSpace(left, right) {
```
While we're in this space, could we improve the parameter names? Names like `tokenBeforeClosingParen` and `closingParenToken` would be much easier to read than `left` and `right`.

platinumazure commented on tests/lib/rules/space-in-parens.js:
```diff
@@ -115,319 +120,417 @@ ruleTester.run("space-in-parens", rule, {
+        },
+        {
+            code: "foo\n(\nbar )",
+            output: "foo\n(\nbar)", // actual error
```
What is the significance of this comment?

GrayedFox replied:
Seems like a relic I left for myself which slipped through, will remove :zap:

GrayedFox commented:
@platinumazure you are in fact correct, the bug I thought I was fixing was due to my own error in interpreting the rule and exception behavior :man_facepalming:... Sigh, a little bit embarrassed, but I did learn a lot along the way.

I've removed the unneeded function and code and kept the small changes that I think make sense. This is now simply an update to the output linting messages, clarification of the rule description, a small bug fix to the context reporting logic. Main PR description and title updated.

GrayedFox commented on docs/rules/space-in-parens.md:
```diff
@@ -177,6 +182,7 @@ Examples of **correct** code for this rule with the `"never", { "exceptions": ["
 
 foo( (1 + 2) );
 foo( (1 + 2), 1);
+foo(bar() );
```
Actually it is in fact "the opposite of the main option will be enforced"! Although it took me many hours to wrap my head around what exactly this code is doing, it's quite the logical puzzle. Here's a line from the docs: 

_"These exceptions work in the context of the first option. That is, if "always" is set to enforce spacing, then any “exception” will disallow spacing. Conversely, if "never" is set to disallow spacing, then any “exception” will enforce spacing."_

So, in this example from this section:

```
// read the following rule and exception declaration as: "spaces are never allowed, except for parentheses, where they are mandatory"
/*eslint space-in-parens: ["error", "never", { "exceptions": ["()"] }]*/
foo(bar() ); // <-- correct, given above rule and exception
foo(bar()); // <-- error, spaces are mandatory for parentheses within parentheses, there should be a space between the last two closing parentheses here
( (foo(bar() ) ) );
```

Interestingly this is the _exact_ case that I thought was a bug, before I fully understood the rule. The exceptions are for curly braces, parentheses, or square brackets that occur *within* grouping parentheses, when they are directly before/after an opening or closing parenthesis. :face_with_head_bandage:

GrayedFox replied:
Haha ummm, yep, I was so confused I fixed what was actually correct behavior before reverting :wink:

GrayedFox commented on tests/lib/rules/space-in-parens.js:
```diff
@@ -115,319 +114,427 @@ ruleTester.run("space-in-parens", rule, {
         // exceptions
+        {
+            code: "foo()",
+            output: "foo( )",
```
That options/exception declaration could be read as "never allow spaces, except in empty parentheses, in which case enforce spaces". The inversion of the rule could be read as "always enforce spaces inside empty parens"

mdjermanovic commented on docs/rules/space-in-parens.md:
```diff
@@ -12,7 +12,9 @@ var x = (1 + 2) * 3;
-This rule will enforce consistency of spacing directly inside of parentheses, by disallowing or requiring one or more spaces to the right of `(` and to the left of `)`. In either case, `()` will still be allowed.
+This rule will enforce consistent spacing directly inside of parentheses, by disallowing or requiring one or more spaces to the right of `(` and to the left of `)`.
+
+As long as you do not explicitly allow or disallow empty parentheses using the `"empty"` exception , `()` will be allowed.
```
This could be confusing, sounds like the rule can disallow empty parentheses (not just allow/disallow spaces).

Maybe 'As long as you do not explicitly disallow `()`...'

GrayedFox replied:
True. I actually would like to simply change this behavior (which would be a breaking change) so that empty is not a special case and works the same way all other exceptions do, i.e. `always` would reject `foo()`, but that is probably best left to another PR and would also warrant some discussion. Updated docs again to be clearer.

mdjermanovic commented on lib/rules/space-in-parens.js:
```diff
@@ -114,102 +119,95 @@ module.exports = {
-         * @param {Object} left The token before the paren
-         * @param {Object} right The paren token
-         * @returns {boolean} True if the paren should have a space
+         * Determines if an opening paren has a space immediately after it
```
This applies to all 4 similar functions, the description should be a bit more specific, e.g. 'has a disallowed space'.

GrayedFox replied:
Nice, good catch, forgot to update the description there :+1:

GrayedFox replied:
Ah, I thought the problem was with the description of the boolean return variable, which I updated - have updated all descriptions to be consistent, let me know what you think

platinumazure commented:
Is there another issue for the breaking change to resolve the inconsistency?

mdjermanovic commented on docs/rules/space-in-parens.md:
```diff
@@ -87,8 +89,21 @@ var foo = ( 1 + 2 ) * 3;
 
+### Empty Exception
+
+The empty exception is a special use case that currently has a few unintuitive behavioral quirks:
```
I believe it's a good thing to have this small section that clarifies how the rule works with empty parens at the moment. 

But I'd be fine with removing this sentence and leaving just the list. 

'currently' might imply that this is going to be changed in the near future (which isn't planned and accepted at the moment), and 'unintuitive' is subjective.

GrayedFox replied:
Roger that

mdjermanovic commented:
> The previous context/report behavior would sometimes report only the first error encountered within a node due to the token checks being nested inside `else if` blocks. This is a non-breaking change: it simply means all errors will be reported within a node instead of only one - it does not introduce a new error type. The fallout of this change is that there is now at least one interesting case: using the default settings `foo( )` technically contains two errors: that single space is both a rejected opening and closing space, so two errors will be reported despite the fact that removing the one space fixes both errors.

I agree that this is a non-breaking change. It's +1 error when there is already an error.

mdjermanovic commented:
> Is there another issue for the breaking change to resolve the inconsistency?

There is an open issue #12141, for several rules with similar behavior.

platinumazure commented:
Merged! Thanks @GrayedFox for sticking with this and being so responsive to feedback!

GrayedFox commented:
@platinumazure @mdjermanovic my pleasure! happy to both receive and respond to feedback, that's what keeps the ship running squeaky clean in the first place :dancer:
